prov/efa/src/rdm/efa_rdm_ep_fiops.c

```diff
@@ -1358,38 +1358,56 @@ static int efa_rdm_ep_getopt(fid_t fid, int level, int optname, void *optval,
 
 	switch (optname) {
 	case FI_OPT_MIN_MULTI_RECV:
+		if (*optlen < sizeof(size_t))
+			return -FI_ETOOSMALL;
```

Comment:
This looks repetitive. How about adding another stack variable `size_t min_len;` and then
```
case ABC:
    min_len = sizeof(type);
    do_something_else;
    break
...

/* After switch cases */
if (*optlen < min_len)
    return -FI_ETOOSMALL;

...
```

Reply:
we should return early before do_something_else because the application may only allocate that size of memory for the opt_val. In this case we cannot defer the check after min_len is assigned.

Reply:
In theory you can still get min_size by another set of `switch case` at the beginning before the block of do something. But that is even worse in duplication

Reply:
> application may only allocate that size of memory for the opt_val

So what happens if we write more data larger than the allocated size? I think it won't crash, does it?

Reply:
it will cause a crash if `opt_val` has 3 bytes allocated but it call fi_getopt for a opt_name of `int` (4 bytes) size and opt_len as 3. Here we should return early when checking opt_len.

Reply:
even if it's not an allocated buffer.  Let's take the case of an option that writes a int64_t sized value, but we call it this way:

```
int32_t optval
int32_t really_important_variable;
size_t optlen = sizeof(optval);
fi_getopt(...., &optval, &optlen);
```
If you write the int64_t sized data before you checked the input optlen, you probably just overwrote `really_important_variable`, and that's probably bad.

Reply:
Cool that makes sense.

In that regard we can also be more defensive by only writing `MIN(*optlen, sizeof(actual_type))` to `optval`

Reply:
> This looks repetitive.

Very solvable problem :)

```c
#define EFA_RDM_EP_GETOPT_CASE(opt, T, expr)	\
	case opt:				\
		if (*optlen < sizeof (T))	\
			return -FI_ETOOSMALL;	\
		*(T *) optval = (expr); 	\
		*optlen = sizeof (T);		\
		break;

// ...

	switch (optname) {
	EFA_RDM_EP_GETOPT_CASE(FI_OPT_MIN_MULTI_RECV, size_t, efa_rdm_ep->min_multi_recv_size)
	EFA_RDM_EP_GETOPT_CASE(FI_OPT_EFA_RNR_RETRY,  size_t, efa_rdm_ep->base_ep.rnr_retry)
	EFA_RDM_EP_GETOPT_CASE(FI_OPT_FI_HMEM_P2P,    int,    efa_rdm_ep->hmem_p2p_opt)
	// ...
	}
```